optimum/habana/distributed/strategy.py

```diff
@@ -0,0 +1,112 @@
+from abc import abstractmethod
+from typing import Any, List, Mapping
+
+import torch
+import torch.distributed
+from torch import nn
+
+#from optimum.habana.distributed import tp_wrapping
+
+
+class DistributedStrategy:
+    def __init__(self, from_meta=False):
+        self.from_meta = from_meta
+
+    def distribute_module(
+        self, module: nn.Module, final_layers: bool = False
+    ) -> nn.Module:
+        """
+        Optionally a distributed strategy may distribute modules that are not
+        numbered layers
+        """
+        return module
+
+    @abstractmethod
+    def distribute_layer(self, block: nn.Module, layer: int) -> nn.Module:
+        """
+        Distribute each layer as-appropriate
+        """
+        pass
+
+
+class NotDistributed(DistributedStrategy):
+    def __init__(self, from_meta=False):
+        super().__init__(from_meta)
+
+    def distribute_module(
+        self, module: nn.Module, final_layers: bool = False
+    ) -> nn.Module:
+        return module
+
+    def distribute_layer(self, block: nn.Module, layer: int) -> nn.Module:
+        return block
+
+
+NoOpStrategy = NotDistributed()
+
+
+class DeviceMover(nn.Module):
+    def __init__(self, module: nn.Module, device):
+        super().__init__()
+        self.device = device
+        # make this wrapper module behave as if it was the wrapped module.
+        attr = module.__dict__
+        attr["module"] = module.to(device)
+        attr["device"] = device
+        self.__dict__ = attr
+
+    def forward(self, *args, **kwargs):
+        device = self.device
+        args = [
+            arg.to(device) if isinstance(arg, torch.Tensor) else arg for arg in args
+        ]
+        kwargs = {
+            k: (
+                kwargs[k].to(device)
+                if isinstance(kwargs[k], torch.Tensor)
+                else kwargs[k]
+            )
+            for k in kwargs
+        }
+        return self.module(*args, **kwargs)
+
+
+class UniformModelParallelStrategy(DistributedStrategy):
+    def __init__(self, devices: List[int], num_layers: int, from_meta=False):
+        super().__init__(from_meta)
+        num_dev = len(devices)
+        layers_per_dev = num_layers // num_dev
+        remainder = num_layers - (layers_per_dev * num_dev)
+        self.layer_to_device = [0] * num_layers
+        layer_id = 0
+        for dev_idx in range(len(devices)):
+            for i in range(layers_per_dev):
+                self.layer_to_device[layer_id] = devices[dev_idx]
+                layer_id = layer_id + 1
+            if remainder > 0:
+                self.layer_to_device[layer_id] = devices[dev_idx]
+                layer_id = layer_id + 1
+                remainder -= 1
+
+    def distribute_layer(self, block: nn.Module, layer: int) -> nn.Module:
+        device = self.layer_to_device[layer]
+        if self.from_meta:
+            # https://github.com/pytorch/pytorch/pull/113647
```

Comment:
this PR is closed, and we can possibly remove the reference to such comments from foundation repo, #Comment

Reply:
Done